### DISCUSSION TITLE
Byte order for this PylontechCANBatteryLimitsFrameStruct

### DISCUSSION THREAD

**Discussion by paulsteigel** [ArminJo/JK-BMSToPylontechCAN#22] [General]
Hi Armin,
It's Ngoc again, I have check your struct on Pylontech_CAN.h
`struct PylontechCANBatteryLimitsFrameStruct {
    struct PylontechCANFrameInfoStruct PylontechCANFrameInfo = { PYLON_CAN_BATTERY_LIMITS_FRAME_ID, 8 }; // 0x351
    struct {
        int16_t BatteryChargeOvervoltage100Millivolt;       // 0 to 750
        int16_t BatteryChargeCurrentLimit100Milliampere;    // 0 to 5000
        int16_t BatteryDischargeCurrentLimit100Milliampere; // -5000 to 0
        int16_t BatteryDischarge100Millivolt;               // 0 to 65535 // not in documentation
    } FrameData;
`
and referring to this https://www.photovoltaikforum.com/core/attachment/77135-pylon-bms-protocol-can-can-20161103-pdf/
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/916aa683-3620-4037-8127-fc50de4548bd)
According to the datasheet, Data0,1 should be charging voltage and Data 3,4 are for current. Following your code, I could not quite understand and I would believe you would go with charge voltage, current then Discharge voltage, current. Am I wrong, **please advice.**
I revised this code to check and start linear charging at 80% SOC (or above) using voltage. Can you advice me on this code with thanks.

```
struct PylontechCANBatteryLimitsFrameStruct {
    struct PylontechCANFrameInfoStruct PylontechCANFrameInfo = { PYLON_CAN_BATTERY_LIMITS_FRAME_ID, 8 }; // 0x351
    struct {
        int16_t BatteryChargeOvervoltage100Millivolt;       // 0 to 750   - for controling charging voltage
        int16_t BatteryChargeCurrentLimit100Milliampere;    // 0 to 5000  - for controlling charging current
        int16_t BatteryDischargeCurrentLimit100Milliampere; // -5000 to 0 - for controlling discharge curent
        int16_t BatteryDischarge100Millivolt;               // 0 to 65535 // not in documentation - for controlling discharging voltage
    } FrameData;
    void fillFrame(struct JKReplyStruct *aJKFAllReply) {
        if (IsCharging && sJKFAllReplyPointer->SOCPercent >= MAX_CURRENT_MODIFICATION_LOWER_SOC_THRESHOLD_PERCENT){
          FrameData.BatteryChargeOvervoltage100Millivolt = map(sJKFAllReplyPointer->SOCPercent,
                                                              MAX_CURRENT_MODIFICATION_LOWER_SOC_THRESHOLD_PERCENT, 100,
                                                              JKComputedData.BatteryFullVoltage10Millivolt / 10,
                                                              JKComputedData.BatteryVoltage10Millivolt/10);
        } else {
          FrameData.BatteryChargeOvervoltage100Millivolt = JKComputedData.BatteryFullVoltage10Millivolt / 10;
          FrameData.BatteryChargeCurrentLimit100Milliampere = swap(aJKFAllReply->ChargeOvercurrentProtectionAmpere) * 10;
          FrameData.BatteryDischargeCurrentLimit100Milliampere = swap(aJKFAllReply->DischargeOvercurrentProtectionAmpere) * 10;
          FrameData.BatteryDischarge100Millivolt = swap(aJKFAllReply->BatteryUndervoltageProtection10Millivolt) / 10;
        }
    }
};
```
The function above will start to gradually reduce charging voltage to the floating level following the increasing of SOC.

I tried your code with modifyCANData and it helped to avoid additional charging at 100% but it will tell the Inverter to update the current limit to a number of other than defined in JKBMS app (which is not quite nice).
I added with IsCharging variable to detect charging status at the getCurrent procedure as below:

```
int16_t getCurrent(uint16_t aJKRAWCurrent) {
    uint16_t tCurrent = swap(aJKRAWCurrent);
    if (tCurrent == 0 || (tCurrent & 0x8000) == 0x8000) {
        // Charge - NO two's complement!
        if (!IsCharging) {
          IsCharging = true;
        }
        return (tCurrent & 0x7FFF);
    }
    // discharge
    if (IsCharging) {
      IsCharging = false;
    }
    return tCurrent * -1;

}
```

Many thanks for your attention,
Ngoc

**Comment by ArminJo** [comment]
> I would believe you would go with charge voltage, current then Discharge voltage, current. 

You are right, but what is wrong with this?

**Reply by paulsteigel** [reply]
Hi Armin,
Thanks for your quick response, comparing with this documentation of the protocol, the program's CAN message might be incorrect rite or the last two byte will not be used (As the CAN ID uses only 6 bytes of data)? Forgive me if I understood your code wrong.
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/e6b46c11-bf09-4f1c-bcaf-cfcd2695153b)

Best!
Ngoc
